sdks/agent-sdk/src/core/Agent.test.ts

```diff
@@ -0,0 +1,151 @@
+import type { AgentContext } from "./AgentContext";
+
+describe("Agent", () => {
+  const mockClient = {
+    inboxId: "test-inbox-id",
+    conversations: {
+      sync: vi.fn().mockResolvedValue(undefined),
+      streamAllMessages: vi.fn(),
+      getConversationById: vi.fn(),
+    },
+    preferences: {
+      inboxStateFromInboxIds: vi.fn(),
+    },
+  };
+
+  const mockConversation = {
+    send: vi.fn().mockResolvedValue(undefined),
+  };
+
+  const mockMessage = {
+    id: "message-id-1",
+    conversationId: "test-conversation-id",
+    senderInboxId: "sender-inbox-id",
+    contentType: ContentTypeText,
+    content: "Hello, world!",
+  } as unknown as DecodedMessage;
+
+  let agent: Agent<unknown>;
+  let options: AgentOptions<unknown>;
+
+  beforeEach(() => {
+    vi.clearAllMocks();
+    options = { client: mockClient as unknown as Client };
+    agent = new Agent(options);
+  });
+
+  describe("types", async () => {
+    const user = createUser();
+    const signer = createSigner(user);
+    const ephemeralAgent = await Agent.create(signer, {
+      env: "dev",
+      dbPath: null,
+      codecs: [new ReplyCodec()],
+    });
+
+    it("infers additional content types from given codecs", () => {
+      expectTypeOf(ephemeralAgent).toEqualTypeOf<
+        Agent<string | Reply | GroupUpdated>
+      >();
+    });
+
+    it("types the content in message event listener", () => {
+      ephemeralAgent.on("message", (ctx) => {
+        expectTypeOf(ctx).toEqualTypeOf<
+          AgentContext<string | Reply | GroupUpdated>
+        >();
+      });
+    });
+  });
+
+  describe("start", () => {
+    it("should sync conversations and start listening", async () => {
+      const mockStream = {
+        [Symbol.asyncIterator]: vi.fn().mockReturnValue({
+          next: vi.fn().mockResolvedValueOnce({ done: true }),
+        }),
+      };
+
+      mockClient.conversations.streamAllMessages.mockResolvedValue(mockStream);
+
+      const startSpy = vi.fn();
+      agent.on("start", startSpy);
+
+      await agent.start();
+
+      expect(mockClient.conversations.streamAllMessages).toHaveBeenCalled();
+      expect(startSpy).toHaveBeenCalled();
+    });
+
+    it("should not start twice if already listening", async () => {
+      const mockStream = {
+        [Symbol.asyncIterator]: vi.fn().mockReturnValue({
+          next: vi.fn().mockResolvedValueOnce({ done: true }),
+        }),
+      };
+
+      mockClient.conversations.streamAllMessages.mockResolvedValue(mockStream);
+
+      const startSpy = vi.fn();
+      agent.on("start", startSpy);
+
+      await agent.start();
+      await agent.start(); // second call should return early
+
+      expect(startSpy).toHaveBeenCalledTimes(1);
+    });
+  });
+
+  describe("use", () => {
+    it("should add middleware and return the agent instance", () => {
+      const middleware = vi.fn();
+
+      const result = agent.use(middleware);
+
+      expect(result).toBe(agent);
+    });
+  });
```

Comment:
should we add tests to ensure that all added middleware is called properly?

Reply:
Good idea, I just added that. Also testing that middleware runs in order.